### DISCUSSION TITLE
openssl版本问题

### DISCUSSION THREAD

**Discussion by qazwsxedckll** [xmake-io/xmake#2228] [Q&A]
我的项目使用了libcurl，依赖openssl1.1.1m，但是项目里有旧的代码，依赖openssl1.0.2u，这种情况怎么解决
add_requires("libcurl", "openssl 1.0.2u")
target("xxx")
    add_packages("libcurl", "openssl")

这样会安装openssl#1 1.1.1m [from:libcurl]
项目编译不过error: ‘HMAC_CTX_init’ was not declared in this scope; did you mean ‘HMAC_CTX_new’?

**Comment by waruqi** [comment]
在一个目标程序里面同时使用，只能使用其中一个版本，可以通过下面的方式改写 curl 依赖的 openssl 版本。。至于用哪个版本你们自己决定

```lua
add_requireconfs("libcurl.openssl", {override = true, version = "1.0.2u"})
```

**Reply by qazwsxedckll** [reply]
手动编译的时候，在编译curl的时候可以选择openssl版本
cd curl-7.80.0 &&./configure --with-openssl=/path_to/openssl-1.1.1l && make && make install
这样就不会影响后面的link

**Reply by qazwsxedckll** [reply]
不好意思再问一下，微服务项目，用的单一仓库，需要生成多个目标程序。有些服务需要旧版本的openssl，有些需要新的，这种情况怎么办。如果用includes的话，是不是版本只能用一个了，没看到能在target里面指定版本。需要每个服务写不同的add_requires，手动执行xmake吗。

**Reply by waruqi** [reply]
```lua
add_requires("openssl~v1 1.1.1l")
add_requires("openssl~v2 1.1.1m")

target("test1")
    add_packages("openssl~v1")
target("test2")
    add_packages("openssl~v2")
```
